src/stirling/source_connectors/socket_tracer/bcc_bpf/grpc_c_trace.c

```diff
@@ -0,0 +1,1557 @@
+    }
+
+    if (0 != get_fd_from_transport((grpc_chttp2_transport *)transport_ptr, &fd))
+    {
+        return -1;
+    }
+
+    // Get the connection info.
+    struct conn_info_t* conn_info = get_or_create_conn_info(pid, fd);
+    if (NULL == conn_info)
+    {
+        return -1;
+    }
+    write_data->conn_id = conn_info->conn_id;
+
+    // Get the headers. They're optional and can be a null pointer.
+    if (0 != get_send_initial_metadata_batch_from_stream(
+            (grpc_chttp2_stream *)stream_ptr,
+            &initial_metadata,
+            version))
+    {
+        return -1;
+    }
+    if (NULL != initial_metadata)
+    {
+        metadata = initiate_empty_grpc_metadata();
+        if (NULL == metadata)
+        {
+            return -1;
+        }
+
+        if (0 != fill_metadata_from_mdelem_list((grpc_mdelem_list *)initial_metadata, metadata))
+        {
+            return -1;
+        }
+
+        if (0 != fire_metadata_events(
+            metadata,
+            write_data->conn_id,
+            write_data->stream_id,
+            write_data->timestamp,
+            write_data->stack_id,
+            write_data->direction,
+            ctx))
+        {
+            return -1;
+        }
+    }
+
+    // Get the trailing headers (AKA trailers). They're optional and can be a null pointer.
+    if (0 != get_send_trailing_metadata_batch_from_stream(
+            (grpc_chttp2_stream *)stream_ptr,
+            &trailing_metadata,
+            version))
+    {
+        return -1;
+    }
+    if (NULL != trailing_metadata)
+    {
+        metadata = initiate_empty_grpc_metadata();
+        if (NULL == metadata)
+        {
+            return -1;
+        }
+
+        if (0 != fill_metadata_from_mdelem_list((grpc_mdelem_list *)trailing_metadata, metadata))
+        {
+            return -1;
+        }
+
+        if (0 != fire_metadata_events(
+            metadata,
+            write_data->conn_id,
+            write_data->stream_id,
+            write_data->timestamp,
+            write_data->stack_id,
+            write_data->direction,
+            ctx))
+        {
+            return -1;
+        }
+    }
+
+    // Get the data.
+    // This only works for uncompressed data or for newer versions (in 1.44 only the flow_controlled_buffer
+    // exists, but in 1.41.1 there's another buffer - compressed_data_buffer).
+    grpc_slice_buffer * flow_controlled_buffer = NULL;
+    if (0 != get_flow_controlled_buffer_from_stream(
+        (grpc_chttp2_stream *)stream_ptr,
+        &flow_controlled_buffer,
+        version))
+    {
+        return -1;
+    }
+
+    u32 total_write_data_length = 0;
+    if (0 != get_slices_from_grpc_slice_buffer_and_fire_perf_event_per_slice(
+        flow_controlled_buffer,
+        write_data,
+        conn_info,
+        ctx))
+    {
+        return -1;
+    }
+
+    return 0;
+}
+
+int probe_grpc_chttp2_mark_stream_closed(struct pt_regs *ctx)
+{
+    struct grpc_c_stream_closed_data data = { 0 };
```

Comment:
nit: Use = {} directly, it is equivalent to {0}.

Reply:
Cool, didn't know that.